source/common/http/http2/codec_impl.h

```diff
@@ -341,13 +352,43 @@ class ConnectionImpl : public virtual Connection,
     bool pending_receive_buffer_high_watermark_called_ : 1;
     bool pending_send_buffer_high_watermark_called_ : 1;
     bool reset_due_to_messaging_error_ : 1;
+    bool defer_processing_backedup_streams_ : 1;
     absl::string_view details_;
 
+    /**
+     * Tracks buffering that may occur for a stream if it is backed up.
+     */
+    struct BufferedStreamManager {
+      bool body_buffered_{false};
+      bool trailers_buffered_{false};
+
+      // We received a call to onStreamClose for the stream, but deferred it
+      // as the stream had pending data to process and the stream was not reset.
+      bool buffered_on_stream_close_{false};
+
+      bool hasBufferedData() const { return body_buffered_ || trailers_buffered_; }
+    };
+
+    BufferedStreamManager stream_manager_;
+    Event::SchedulableCallbackPtr process_buffered_data_callback_;
+
   protected:
     // Http::MultiplexedStreamImplBase
     bool hasPendingData() override {
       return pending_send_data_->length() > 0 || pending_trailers_to_encode_ != nullptr;
     }
+    bool continueProcessingBufferedData() const {
+      // We should stop processing buffered data if either
+      // 1) Buffers become overrun
+      // 2) The stream ends up getting reset
+      // Both of these can end up changing as a result of processing buffered data.
+      return !buffersOverrun() && !reset_reason_.has_value();
+    }
+
+    // Avoid inversion in the case where we saw trailers, acquiring the
+    // remote_end_stream_ being set to true, but the trailers ended up being
+    // buffered.
+    bool sendEndStream() const { return remote_end_stream_ && !stream_manager_.trailers_buffered_; }
```

Comment:
What about !stream_manager_.body_buffered_ ?

This seems relevant in ConnectionImpl::ClientStreamImpl::decodeHeaders(), the other 2 calls I can understand at least while you're moving the full body through the filter pipeline in 1 go.

Reply:
Worth an ASSERT(!stream_manager_.body_buffered_) ?